### PR TITLE
feat: impl show components

### LINKED ISSUE #1232: feat: support sql `SHOW COMPONENTS`
`SHOW COMPONENTS` prints all components registered in the cluster, which include 
- endpoint
- role: tablet/name server/api server/task manager
- connect_time
- status : healthy or offline

for tablet there is healthy/offline status, for name server /api server, there is jut healthy, offline components does not exists in the list

### PR DESCRIPTION
**Pull request by aceforeverd**
close #1232

A correct example:
Screen Shot 2022-03-11 at 10 59 46 AM

### CODE REVIEW

File: src/sdk/sql_cluster_router.cc [hunk 1]
```diff
@@ -2960,5 +2990,189 @@ bool SQLClusterRouter::CheckPreAggrTableExist(const std::string& base_table, con
...
+        data[i][3] = std::to_string(it->second);
+
+        // status
+        // offlined nameserver won't register in zookeeper, so there is only online
+        data[i][4] = "online";
+
+        // ns_role
+        // NSs runs as mater/standby mode
+        if (i == 0) {
+            data[i][5] = "leader";
```
Comment (aceforeverd) [inline src/sdk/sql_cluster_router.cc]: master

File: src/sdk/sql_cluster_router.cc [hunk 2]
```diff
@@ -2960,5 +2990,189 @@ bool SQLClusterRouter::CheckPreAggrTableExist(const std::string& base_table, con
...
+        Stat stat;
+        if (!zk_client->GetNodeValueAndStat(real_path.c_str(), &endpoint, &stat)) {
+            status->code = hybridse::common::kRunError;
+            status->msg = absl::StrCat("get endpoint failed for path: ", real_path);
+            return {};
+        }
+        if (endpoint_set.find(endpoint) == endpoint_set.end()) {
+            endpoint_set[endpoint] = stat.ctime;
+        } else {
+            endpoint_set[endpoint] = std::min(stat.ctime, endpoint_set[endpoint]);
```
Comment (aceforeverd) [inline src/sdk/sql_cluster_router.cc]: `ctime` seems inaccurate.
Capture

**Comment by aceforeverd** [comment]
a offline node's connect time do not reset to zero:
Capture

File: src/sdk/sql_cluster_router.cc [hunk 3]
```diff
@@ -1405,7 +1441,8 @@ std::shared_ptr<hybridse::sdk::ResultSet> SQLClusterRouter::HandleSQLCmd(
                 return {};
             }
             auto tables = cluster_sdk_->GetTables(db);
-            std::vector<std::vector<std::string>> values;;
+            std::vector<std::vector<std::string>> values;
+            ;
```
Comment (vagetablechicken) [inline src/sdk/sql_cluster_router.cc]: only one semicolon?

**Comment by vagetablechicken** [comment]
> a offline node's connect time do not reset to zero: Capture

Can we make the `connect time` more readable? `std::gmtime` or other way?

**Comment by aceforeverd** [comment]
> > a offline node's connect time do not reset to zero: Capture
> 
> Can we make the `connect time` more readable? `std::gmtime` or other way?

the prometheus will going to parse the time, so I guess we should do some hack in the cli ?

**Comment by vagetablechicken** [comment]
> > > a offline node's connect time do not reset to zero: Capture
> > 
> > 
> > Can we make the `connect time` more readable? `std::gmtime` or other way?
> 
> the prometheus will going to parse the time, so I guess we should do some hack in the cli ?

Yes, just print, no need to change the type.

**Comment by aceforeverd** [comment]
> > > > a offline node's connect time do not reset to zero: Capture
> > > 
> > > 
> > > Can we make the `connect time` more readable? `std::gmtime` or other way?
> > 
> > 
> > the prometheus will going to parse the time, so I guess we should do some hack in the cli ?
> 
> Yes, just print, no need to change the type.

Well I may misunderstood. `CONNECT_TIME` here is actually impled as `connected_time` (time from now to first  connect time), which is a time interval. I think it might better to impl it as `CONNECT_TIME` , which is the timestamp. For timestamp, we already pretty-printed in the console

**Comment by vagetablechicken** [comment]
So the `connected_time` is just xx microseconds interval?  Shall we convert it to xx min or hour

**Comment by aceforeverd** [comment]
> So the `connected_time` is just xx microseconds interval? Shall we convert it to xx min or hour

Yes, now I fixed it and it is now regarded as timestamp. The `ResultSetSQL::MakeResultSet` do not support put a timestamp or int into it directly, so I'll fix pretty-print issue later

**Review by vagetablechicken — APPROVED** [review]

**Review by dl239 — APPROVED** [review]

**Pull request merged**